core/src/main/scala/filodb.core/memstore/PartKeyLuceneIndex.scala

```diff
@@ -289,18 +300,39 @@ class PartKeyLuceneIndex(ref: DatasetRef,
     indexWriter.close()
   }
 
-  // TODO - most frequently used eviction needed
-  private val readerStateCache = mutable.WeakHashMap[(IndexReader, String), DefaultSortedSetDocValuesReaderState]()
+  // DefaultSortedSetDocValuesReaderState (per label/reader) is expensive to create, so cache & reuse across queries.
+  // Different cache for shard keys and non-shard-keys since it is evicted differently.
+  // TODO cache config is currently hardcoded - make configuration properties if really deemed necessary
+  private val readerStateCacheShardKeys: LoadingCache[(IndexReader, String), DefaultSortedSetDocValuesReaderState] =
+  Caffeine.newBuilder()
+    .maximumSize(100)
+    .recordStats()
+    .build((key: (IndexReader, String)) => {
+      new DefaultSortedSetDocValuesReaderState(key._1, FACET_FIELD_PREFIX + key._2)
+    })
+  private val readerStateCacheNonShardKeys: LoadingCache[(IndexReader, String), DefaultSortedSetDocValuesReaderState] =
+    Caffeine.newBuilder()
+      .maximumSize(200)
+      .recordStats()
+      .build((key: (IndexReader, String)) => {
+        new DefaultSortedSetDocValuesReaderState(key._1, FACET_FIELD_PREFIX + key._2)
+      })
+
   def labelValuesEfficient(colFilters: Seq[ColumnFilter], startTime: Long, endTime: Long,
                            colName: String, limit: Int = 100): Seq[String] = {
-    require(facetEnabled, "Faceting not enabled on this index; labelValuesEfficient cannot be called")
+    require(facetEnabledAllLabels ||
+      facetEnabledSharedKeyLabels && schema.options.shardKeyColumns.contains(colName),
+      s"Faceting not enabled for label $colName; labelValuesEfficient should not have been called")
```

Comment:
This is same as ``require(facetEnabledForLabel(colName), ...)``

Reply:
done